src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -258,39 +274,48 @@ private void HandleChannelException(IConnection connection, Exception ex)
                 return;
             }
 
-            if (ShouldInvalidateServer(ex))
+            if (connection.Generation != _connectionPool.Generation)
```

Comment:
This check comes from this line in the spec commit
https://github.com/mongodb/specifications/commit/7dd9c008be6edaf65368787d5e91c2ac3aeb95b8#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1217-R1223
```
When processing a network or command error, clients MUST first check the
error's `generation number`_. If the error's generation number is equal to
the pool's generation number then error handling MUST continue according to
`Network error when reading or writing`_ or
`"not master" and "node is recovering"`_. Otherwise, the error is considered
stale and the client MUST NOT update any topology state.
(See `Why ignore errors based on CMAP's generation number?`_)
```